src/v/utils/waiter_queue.h

```diff
@@ -51,6 +52,7 @@ class waiter_queue {
         auto sub_opt = as.subscribe(
           [item_ptr = item->weak_from_this()]() noexcept {
               if (item_ptr) {
+                  item_ptr->aborted = true;
```

Comment:
what about removing the item from the _items set? in some ways this makes for an awkward api because if a waiter is aborted, it looks like we have to depend on notify to be called to gc the items?

Reply:
yes, i assumed that the item will be removed when next notify is called. Capturing `this` in the callback isn't great either. @jcsp what do you think ?

Reply:
Yeh, good point. I think if you add a destructor then you could make it safe to capture this so there weren't any active callbacks after destruction finished.